### PR TITLE
User geo information endpoint

### PR DESCRIPTION
**Pull request by hero101**
### Describe the background of your pull request

What does your pull request do? Does it solve a bug (which one?), add a feature?

### Additional context

Add any other context

### Governance 

- [ ] Documentation is added
- [ ] Test cases are added or updated

By submitting this pull request I confirm that:
- I've read the contributing document https://github.com/alkem-io/.github/blob/master/CONTRIBUTING.md.
- I've read and understand the Code of Conduct https://github.com/alkem-io/.github/blob/master/CODE_OF_CONDUCT.md.
- I understand that any contributions or suggestions I made may make it into the actual code. I've read the License 
     https://github.com/alkem-io/Coordination/blob/master/LICENSE.
- I understand that submitting the pull request requires agreeing to and signing the contributor license agreement.

### CODE REVIEW

File: src/app.controller.ts [hunk 1]
```diff
@@ -46,4 +65,37 @@ export class AppController {
...
+    @Req() req: Request
+  ): Promise<GeoInformation | undefined> {
+    const forwardedFor = req.headers['x-forwarded-for'];
+
+    if (!forwardedFor) {
+      this.logger.error('X-Forwarded-For header not set');
+      return undefined;
+    }
+
+    const ip = String(forwardedFor).split(',')?.[0];
```
Comment (valentinyanakiev) [inline src/app.controller.ts:83]: Are you sure it's the first and not the last item in the array?

File: src/common/enums/alkemio.error.status.ts [hunk 2]
```diff
@@ -58,4 +58,7 @@ export enum AlkemioErrorStatus {
   SESSION_EXTEND = 'SESSION_EXTEND',
   BEARER_TOKEN = 'BEARER_TOKEN',
   CALLOUT_CLOSED = 'CALLOUT_CLOSED',
+  GEO_SERVICE_NOT_AVAILABLE = 'GEO_SERVICE_NOT_AVAILABLE',
+  GEO_SERVICE_ERROR = 'GEO_SERVICE_ERROR',
+  GEO_SERVICE_LIMIT_REACHED = 'GEO_SERVICE_LIMIT_REACHED',
```
Comment (valentinyanakiev) [inline src/common/enums/alkemio.error.status.ts]: It is a standard notation to name the exception with `exceeded`. There exception occurs when you exceed the limit, not when you reach it.
```suggestion
  GEO_SERVICE_REQUEST_LIMIT_EXCEEDED = 'GEO_SERVICE_REQUEST_LIMIT_EXCEEDED',
```

File: src/common/exceptions/geo/geo.service.limit.reached.exception.ts [hunk 3]
```diff
@@ -0,0 +1,8 @@
+import { BaseException } from '@common/exceptions/base.exception';
+import { AlkemioErrorStatus, LogContext } from '@common/enums';
+
+export class GeoServiceLimitReachedException extends BaseException {
+  constructor(error: string, context: LogContext) {
+    super(error, context, AlkemioErrorStatus.GEO_SERVICE_LIMIT_REACHED);
```
Comment (valentinyanakiev) [inline src/common/exceptions/geo/geo.service.limit.reached.exception.ts]:
```suggestion
    super(error, context, AlkemioErrorStatus.GEO_SERVICE_REQUEST_LIMIT_EXCEEDED);
```

File: src/common/exceptions/geo/geo.service.limit.reached.exception.ts [hunk 4]
```diff
@@ -0,0 +1,8 @@
+import { BaseException } from '@common/exceptions/base.exception';
+import { AlkemioErrorStatus, LogContext } from '@common/enums';
+
+export class GeoServiceLimitReachedException extends BaseException {
```
Comment (valentinyanakiev) [inline src/common/exceptions/geo/geo.service.limit.reached.exception.ts]:
```suggestion
export class GeoServiceRequestLimitExceededException extends BaseException {
```

File: src/common/exceptions/geo/index.ts [hunk 5]
```diff
@@ -0,0 +1,3 @@
+export * from './geo.service.not.available.exception';
+export * from './geo.service.error.exception';
+export * from './geo.service.limit.reached.exception';
```
Comment (valentinyanakiev) [inline src/common/exceptions/geo/index.ts]:
```suggestion
export * from './geo.service.limit.exceeded.exception';
```

Also, rename the file.

File: src/services/geo-location/geo.location.service.ts [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+    const userGeoCached = await this.cacheManager.get<GeoInformation>(ip);
+
+    if (userGeoCached) {
+      return userGeoCached;
+    }
+
+    if (!(await this.canCallService())) {
+      throw new GeoServiceLimitReachedException(
+        `3rd party service limit of ${allowedCallsToService} calls per ${allowedCallsToServiceWindow} seconds reached`,
+        LogContext.UNSPECIFIED
```
Comment (valentinyanakiev) [inline src/services/geo-location/geo.location.service.ts]: Add a LogContext and specify it.

File: src/services/geo-location/geo.location.service.ts [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+      .get<GeoPluginResponse>(`${endpoint}${ip}`)
+      .toPromise()
+      .catch((err: AxiosError) => {
+        return err;
+      });
+
+    if (!response) {
+      throw new GeoServiceNotAvailableException(
+        `3rd party service at (${endpoint}) not available`,
+        LogContext.UNSPECIFIED
```
Comment (valentinyanakiev) [inline src/services/geo-location/geo.location.service.ts]: Add log context and specify it.

File: src/services/geo-location/geo.location.service.ts [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+      );
+    }
+
+    if (response instanceof Error) {
+      throw new GeoServiceErrorException(
+        `3rd party service returned an error: ${response.message}`,
+        LogContext.UNSPECIFIED
```
Comment (valentinyanakiev) [inline src/services/geo-location/geo.location.service.ts]: Add log context and specify it.

File: src/services/geo-location/geo.location.service.ts [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+const endpoint = 'http://www.geoplugin.net/json.gp?ip=';
+const cacheEntryTtl = 14400; // 4 hours
+const geoServiceCallsKey = 'geo-service-call-limit';
+// https://www.geoplugin.com/faq#i_stopped_getting_responses_from_geoplugin.net
+// the free lookup limit of 120 requests per minute.
+const allowedCallsToService = 120;
+const allowedCallsToServiceWindow = 60;
+
+@Injectable()
+export class GeoLocationService {
```
Comment (valentinyanakiev) [inline src/services/geo-location/geo.location.service.ts]: This service is one of the things that would make sense to have unit tests, as there is a flow (algorithm) in it.

File: src/services/geo-location/geo.location.service.ts [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+import {
+  GeoServiceErrorException,
+  GeoServiceLimitReachedException,
+  GeoServiceNotAvailableException,
+} from '@common/exceptions/geo';
+import { LogContext } from '@common/enums';
+import { GeoInformation } from './geo.information';
+import { GeoPluginResponse } from './geo.plugin.response';
+
+const endpoint = 'http://www.geoplugin.net/json.gp?ip=';
```
Comment (valentinyanakiev) [inline src/services/geo-location/geo.location.service.ts]: Move these entries to config, it is not impossible for them to change the endpoint or rate limits. We shouldn't need to redeploy then...

File: src/services/geo-location/geo.location.service.ts [hunk 11]
```diff
@@ -0,0 +1,115 @@
+import { CACHE_MANAGER, Inject, Injectable } from '@nestjs/common';
+import { HttpService } from '@nestjs/axios';
+import { AxiosError } from 'axios';
+import { Cache } from 'cache-manager';
+import { GeoLocationCacheMetadata } from '@services/geo-location/geo.location.cache.metadata';
+import {
+  GeoServiceErrorException,
+  GeoServiceLimitReachedException,
```
Comment (valentinyanakiev) [inline src/services/geo-location/geo.location.service.ts]:
```suggestion
  GeoServiceLimitExceededException,
```

File: src/services/geo-location/geo.location.service.ts [hunk 12]
```diff
@@ -0,0 +1,115 @@
...
+
+  public async getGeo(ip: string): Promise<GeoInformation | undefined> {
+    const userGeoCached = await this.cacheManager.get<GeoInformation>(ip);
+
+    if (userGeoCached) {
+      return userGeoCached;
+    }
+
+    if (!(await this.canCallService())) {
+      throw new GeoServiceLimitReachedException(
```
Comment (valentinyanakiev) [inline src/services/geo-location/geo.location.service.ts]:
```suggestion
      throw new GeoServiceLimitExceededException(
```

**Review by valentinyanakiev — CHANGES_REQUESTED** [review]
- Move the geo-location folder from src/services/geo-location to src/services/external/geo-location
- Rename geo.service.limit.reached.exception.ts to geo.service.limit.exceeded.exception.ts

The rest of the comments you can find inline. I will see how to test the code.

File: src/services/geo-location/geo.location.service.ts (hunk 6 above)
Comment (valentinyanakiev) [reply src/services/geo-location/geo.location.service.ts]: LogContext.GEO

File: src/app.controller.ts (hunk 1 above)
Comment (hero101) [reply src/app.controller.ts:83]: That's what I understood from https://stackoverflow.com/a/72586833

Comment (valentinyanakiev) [reply src/app.controller.ts:83]: I think you are right, but we need to check.

**Review by valentinyanakiev — APPROVED** [review]

**Pull request merged**